torch/optim/optimizer.py

```diff
@@ -317,3 +308,20 @@ def add_param_group(self, param_group):
             raise ValueError("some parameters appear in more than one parameter group")
 
         self.param_groups.append(param_group)
+
+    def update_parameters(self, params):
```

Comment:
Hm, I think we have to be very careful introducing this new public API.
* I think the name should be tweaked since applying gradient updates could be thought of as "updating parameters" and that's not what's being done here (maybe `swap_parameters()` or similar is better?).
* This mutating API allows the "parameters remain constant throughout the optimizer lifetime" invariant that used to hold for optimizers to be broken. I wonder if there are areas in the code that assume this invariant and break if it doesn't hold (e.g. LR schedulers)?

For stateful optimizers such as SGD, is this the only way to support optimizing on a different set of params?

Reply:
Also this fully resets the current state of the optimizer.
For the purpose of this PR and the meta learning example we have, I don't think we need such a function.
Can you simply create the optimizer inside the test function?

Reply:
> For stateful optimizers such as SGD, is this the only way to support optimizing on a different set of params?

I don't think of another possible way to support this other than passing a `param_groups` kwarg to `step` to act in a functional way

Reply:
> Can you simply create the optimizer inside the test function?

I think this is fine, but do we want to create an optimizer object for every step?
Since the user will be able to provide all the needed internal `state` values creating two optimizers for subsequent updates should be equivalent to calling `.step` on the same optimizer twice. However, the way that LR schedulers work could be affected?